File: implementations/python/merge_sort/merge_sort_large.py
import sys

N = 1_000_000

arr = [0] * N

def setup():
    pass  # reset handles initialization

def reset():
    for i in range(N):
        arr[i] = N - i

def merge_sort_impl(left, right):
    if left >= right:
        return
    mid = left + (right - left) // 2
    merge_sort_impl(left, mid)
    merge_sort_impl(mid + 1, right)
    merge(left, mid, right)

def merge(left, mid, right):
    temp = arr[left:right + 1]
    i, j, k = 0, mid - left + 1, left
    while i <= mid - left and j <= right - left:
        if temp[i] <= temp[j]:
            arr[k] = temp[i]
            i += 1
        else:
            arr[k] = temp[j]
            j += 1
        k += 1
    while i <= mid - left:
        arr[k] = temp[i]
        i += 1
        k += 1
    while j <= right - left:
        arr[k] = temp[j]
        j += 1
        k += 1

def merge_sort():
    reset()
    merge_sort_impl(0, N - 1)
    return arr[N - 1]

def main():
    setup()

    sys.stdout.write("ready\n")
    sys.stdout.flush()

    for line in sys.stdin:
        sys.stdout.write(str(merge_sort()) + "\n")
        sys.stdout.flush()

if __name__ == "__main__":
    main()

File: implementations/javascript/merge_sort/merge_sort_large.js
const readline = require('readline');

const N = 1000000;
const arr = new Array(N).fill(0);

function setup() {
    // reset handles initialization
}

function reset() {
    for (let i = 0; i < N; i++) {
        arr[i] = N - i;
    }
}

function merge(left, mid, right) {
    const temp = arr.slice(left, right + 1);
    let i = 0, j = mid - left + 1, k = left;
    while (i <= mid - left && j <= right - left) {
        if (temp[i] <= temp[j]) {
            arr[k++] = temp[i++];
        } else {
            arr[k++] = temp[j++];
        }
    }
    while (i <= mid - left) arr[k++] = temp[i++];
    while (j <= right - left) arr[k++] = temp[j++];
}

function merge_sort_impl(left, right) {
    if (left >= right) return;
    const mid = left + Math.floor((right - left) / 2);
    merge_sort_impl(left, mid);
    merge_sort_impl(mid + 1, right);
    merge(left, mid, right);
}

function merge_sort() {
    reset();
    merge_sort_impl(0, N - 1);
    return arr[N - 1];
}

setup();
process.stdout.write("ready\n");

const rl = readline.createInterface({ input: process.stdin });
rl.on('line', () => {
    process.stdout.write(String(merge_sort()) + "\n");
});